File: cpp_intro2algo/dataStructures.cpp
/*--***************************************************************************

Chapter 10: Basic Structures (p.129)
    Queue                   (p.130) 

Chapter 6:
    Priority Queue          (p.90)

***************************************************************************--*/


#include <iostream>
#include "dataStructures.h"
#include "sorting.h"


Queue::Queue(){
    _size = 0;
    _origin = NULL;
    _head = -1;
    _tail = -1;
}

Queue::Queue(int* aList, const int size){
    if (size == 0){
        _size = 0;
    }else{
        _size = size;
        _origin = aList;
        _head = 0;
        _tail = size - 1;
    }
}

bool Queue::is_empty(){
    if (_size == 0){
        return true;
    }else if (_head == _tail){
        return true;
    }else{
        return false;
    }
}

int Queue::get_head(){
    if (not is_empty()){
        return _head;
        //return *(_origin + _head);
    }else{
        return -1;
    }

}

int Queue::get_tail(){
    if (not is_empty()){
        return _tail;
        //return *(_origin + _tail);
    }else{
        return -1;
    }
}

void Queue::showq(){
    if (is_empty()){
        //pass
    }else if (_head <= _tail){
        for (int i=_head; i<=_tail; i++){
            std::cout << *(_origin + i) << " ";
        }
    }else{
        for (int i=_head; i<_size; i++){
            std::cout << *(_origin + i) << " ";
        }
        for (int i=0; i<=_tail; i++){
            std::cout << *(_origin + i) << " ";
        }
    }
    std::cout << std::endl;
}

void Queue::enqueue(int x){
    if (_size == 0){
        std::cout << "Enqueue failed. Queue is full." << std::endl;
    }else if ((_tail + 1 - _head) % _size){
        _tail = (_tail + 1) % _size;
        *(_origin + _tail) = x;
    }else{
        std::cout << "Enqueue failed. Queue " << "is full." << std::endl;
    }
}

int Queue::dequeue(){
    if (is_empty()){
        std::cout << "Dequeue failed. Queue " << "is empty." << std::endl;
    }else{
       _head = (_head + 1) % _size;
       return *(_origin + _head - 1);
    }
}

File: py_intro2algo/dataStructures.py
'''
Basic Structures:
    Queue                   (p.130)

Chapter 6:
    Priority Queue          (p.161)


'''
from __future__ import print_function


class Queue(list):
    def __init__(self, aList):
        list.__init__(self, aList)
        self._size = len(aList)
        self._head = 0
        self._tail = self._size - 1

    def is_empty(self):
        if self._size == 0:
            return True
        elif self._tail + 1 == self._head:
            return True
        else:
            return False

    def get_head(self):
        if self.is_empty():
            return None
        else:
            return self._head

    def get_tail(self):
        if self.is_empty():
            return None
        else:
            return self._tail

    def showq(self):
        if self.is_empty():
            print()
        else:
            h = self._head % self._size
            t = (self._tail + 1) % self._size
            if h < t:
                for x in self[h:t]:
                    print(x, end=" ")
                print()
            else:
                for x in self[h: ]:
                    print(x, end=" ")
                for x in self[ :t]:
                    print(x, end=" ")
                print()

    def enqueue(self, x):
        if self._tail + 1 - self._head == self._size:
            print("Enqueue failed. Queue is full.")
        else:
            if (self._tail + 1 - self._head) % self._size:
                self._tail += 1
                self[self._tail % self._size] = x
            else:
                print("Enqueue dailed. Queue is full.")

    def dequeue(self):
        if self.is_empty():
            print("Dequeue failed. Queue is empty.")
        else:
            self._head += 1
            return self[self._head - 1]
